### PR TITLE
[2.2.6] Fix watcher loop quit and channel shouldDrop logic

### PR DESCRIPTION
**Pull request by congqixia**
Also bump version to 2.2.6
See also #23393 
/kind bug

### CODE REVIEW

File: internal/querynode/shard_node_detector.go [hunk 1]
```diff
@@ -132,31 +133,76 @@ func (nd *etcdShardNodeDetector) watch(ch clientv3.WatchChan, collectionID, repl
...
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	err := retry.Do(ctx, func() error {
+		ch = nd.client.Watch(ctx, nd.path, clientv3.WithPrefix())
+		select {
+		case evt, ok := <-ch:
+			if !ok {
+				return errors.New("rewatch got closed ch")
+			}
+			nd.handleEvt(evt, collectionID, replicaID)
```
Comment (czs007) [inline internal/querynode/shard_node_detector.go:187]: what's the purpose of this handleEvt?

File: internal/querynode/shard_node_detector.go [hunk 2]
```diff
@@ -132,31 +135,93 @@ func (nd *etcdShardNodeDetector) watch(ch clientv3.WatchChan, collectionID, repl
...
+		default:
+			// blocked, fine
+		}
+		return nil
+	})
+	// check detector closed
+	if err != nil {
+		select {
+		case <-nd.closeCh:
+			return nil, false, revision
```
Comment (jaime0815) [inline internal/querynode/shard_node_detector.go:218]: nd.closeCh has already been processed at the 192 lines, and there will never be executed?

File: internal/querynode/shard_node_detector.go [hunk 3]
```diff
@@ -132,31 +135,93 @@ func (nd *etcdShardNodeDetector) watch(ch clientv3.WatchChan, collectionID, repl
...
+		default:
+			panic(err)
```
Comment (jaime0815) [inline internal/querynode/shard_node_detector.go:220]: It seems still panic while return error at the 202 lines

File: internal/querynode/shard_node_detector.go (hunk 2 above)
Comment (congqixia) [reply internal/querynode/shard_node_detector.go:218]: Line 192 handles channel closed in retry function only

File: internal/querynode/shard_node_detector.go (hunk 3 above)
Comment (congqixia) [reply internal/querynode/shard_node_detector.go:220]: this line panics when retry (default 10 times) failed to rewatch

File: internal/querynode/shard_node_detector.go [hunk 4]
```diff
@@ -132,31 +135,93 @@ func (nd *etcdShardNodeDetector) watch(ch clientv3.WatchChan, collectionID, repl
...
+			ch = nd.client.Watch(ctx, nd.path, clientv3.WithPrefix(), clientv3.WithRev(revision+1))
+		} else {
+			ch = nd.client.Watch(ctx, nd.path, clientv3.WithPrefix())
+		}
+		select {
+		case <-nd.closeCh:
+			return retry.Unrecoverable(errors.New("detector closed"))
+
+		case evt, ok := <-ch:
+			if !ok {
```
Comment (jaime0815) [inline internal/querynode/shard_node_detector.go:196]: We'd better check the minimum CompactRevision value from WatchChan and ensure the watched revision is not compacted.

Comment (jaime0815) [reply internal/querynode/shard_node_detector.go:196]: if watched revision is compacted, the client will receive ErrCompacted

Comment (congqixia) [reply internal/querynode/shard_node_detector.go:196]: If watched revision is compacted, a response with `ErrCompacted` will be received. Any internally, ETCD client will update the latest revision for each event received.

File: internal/datacoord/channel_manager.go [hunk 5]
```diff
@@ -674,21 +675,25 @@ func (c *ChannelManager) watchChannelStatesLoop(ctx context.Context) {
...
 				return
 			}
 
 			if err := event.Err(); err != nil {
 				log.Warn("datacoord watch channel hit error", zap.Error(event.Err()))
 				// https://github.com/etcd-io/etcd/issues/8980
+				// use latest
 				if event.Err() == v3rpc.ErrCompacted {
-					go c.watchChannelStatesLoop(ctx)
+					go c.watchChannelStatesLoop(ctx, common.LatestRevision)
```
Comment (czs007) [inline internal/datacoord/channel_manager.go:688]: should be version + 1?

**Pull request merged**